src/main/java/com/exasol/spark/s3/ExasolWriteBuilderProvider.java

```diff
@@ -0,0 +1,112 @@
+     */
+    public WriteBuilder createWriteBuilder(final StructType schema, final LogicalWriteInfo defaultInfo) {
+        final SparkSession sparkSession = SparkSession.active();
+        final LogicalWriteInfo info = getUpdatedLogicalWriteInfo(defaultInfo, sparkSession);
+        final ExasolOptions updatedOptions = getUpdatedOptions(info.options());
+        final String intermediateDataPath = updatedOptions.get(INTERMEDIATE_DATA_PATH);
+        LOGGER.info(() -> "Writing intermediate data to the '" + intermediateDataPath + "' path for write job.");
+        final CSVTable csvTable = new CSVTable("", sparkSession, info.options(), getS3WritePath(intermediateDataPath),
+                Option.apply(schema), new CSVFileFormat().getClass());
+        return new DelegatingWriteBuilder(updatedOptions, csvTable.newWriteBuilder(info));
+    }
+
+    private ExasolOptions getUpdatedOptions(final Map<String, String> map) {
+        final ExasolOptions.Builder builder = ExasolOptions.builder() //
+                .jdbcUrl(this.options.getJdbcUrl()) //
+                .username(this.options.getUsername()) //
+                .password(this.options.getPassword()) //
+                .s3Bucket(this.options.getS3Bucket());
+        if (this.options.hasTable()) {
+            builder.table(this.options.getTable());
+        } else {
+            builder.query(this.options.getQuery());
+        }
+        builder.withOptionsMap(map);
+        return builder.build();
+    }
+
+    private LogicalWriteInfo getUpdatedLogicalWriteInfo(final LogicalWriteInfo defaultInfo,
+            final SparkSession sparkSession) {
+        final Map<String, String> map = new HashMap<>(defaultInfo.options().asCaseSensitiveMap());
+        map.put("header", "true");
+        map.put("delimiter", ",");
+        map.put("mapreduce.fileoutputcommitter.marksuccessfuljobs", "false");
+        final String s3Bucket = this.options.getS3Bucket();
+        final String s3BucketKey = getS3BucketKeyForWriteLocation(defaultInfo, sparkSession);
+        map.put(INTERMEDIATE_DATA_PATH, "s3a://" + Paths.get(s3Bucket, s3BucketKey).toString());
+        map.put(WRITE_S3_BUCKET_KEY, s3BucketKey);
+
+        return new LogicalWriteInfo() {
+            @Override
+            public String queryId() {
+                return defaultInfo.queryId();
+            }
+
+            @Override
+            public StructType schema() {
+                return defaultInfo.schema();
+            }
+
+            @Override
+            public CaseInsensitiveStringMap options() {
+                return new CaseInsensitiveStringMap(map);
+            }
+        };
+    }
+
+    private String getS3BucketKeyForWriteLocation(final LogicalWriteInfo defaultInfo, final SparkSession sparkSession) {
+        final StringBuilder builder = new StringBuilder();
+        builder.append(UUID.randomUUID()).append("-").append(sparkSession.sparkContext().applicationId()).append("/")
```

Comment:
It should be fine. The generated UUID is combined with sparkContext applicationId (this is also unique id) as a bucketKey. 

The application id will be unique per each spark job, for example, submitted using `spark-submit`. In the `spark-shell` there is single context, but this is also fine since on interactive environment only one Spark application will be running at a time.

Reply:
ok, maybe an additional question, what happens when the directory already exists (unlikely, I know)?

Reply:
Create #164 for it